### PR TITLE
Atomic Components

### PR DESCRIPTION
**Pull request by stropho**
Based on and hopefully 
Closes https://github.com/parsable/react-truncate-markup/issues/48

#### Well, I also added a few changes not directly connected to the mentioned issue above:
1. upgrade react peer dependency to 16.3 because of `UNSAFE_componentWillRecieveProps` - mentioned in https://github.com/parsable/react-truncate-markup/issues/29
2. in my effort to eliminate code in `UNSAFE_componentWillRecieveProps` , some calculations from there were moved to memoized getters, using `memoize-one`. I think the performance actually  benefits, on the other hand it adds another dependency. Totally opened for a discussion.
3. validation didn't correctly process children, which could be arrays of arrays - fixed, tested

#### And finally

4. `Atom` component available via `<TruncateMarkup.Atom />`, including
 - updating warning message, unit tests
 - example in demo page

The changes are split to multiple commits, it's probably easier to see them one by one :crossed_fingers: 

Btw. in the sourcecode, there is a mention about `EC#1` ? I have no idea what that means neither if it's something the new implementation should also take care of.

### CODE REVIEW

**Comment by stropho** [comment]
ping @patrik-piskay

**Comment by patrik-piskay** [comment]
Hi @stropho, sorry for the slow progress on this. Overall I think this PR is good but I'll need more time to properly test it and I'll make some changes to the code/examples/readme before merging. , I'll have some more time to work on this from next week. 🤞

File: src/index.js [hunk 1]
```diff
@@ -130,51 +128,60 @@ export default class TruncateMarkup extends React.Component {
...
   componentDidMount() {
-    if (!this.isValid) {
+    if (!this.isValid()) {
       return;
     }
-
-    this.origText = this.state.text;
-
+    // trigger warning, if any
+    this.getPolicy();
```
[the remark this answers is not in the document]
Comment (stropho) [reply src/index.js:165]: Totally agree. I think it required changing an existing test for the warning. But maybe I am mistaken. I'll look into it this week .

**Comment by stropho** [comment]
> but when resizing its box it never recovers
I actually saw the same issue even with characters. Here I thought: what a smart idea to set minimum width for the for the resizebox in the examples :smile:

**Comment by patrik-piskay** [comment]
Hi @stropho, I'd like to push changes to this PR, would you mind giving me permissions to do so? [There should be a checkbox in the PR view](https://github.blog/2016-09-07-improving-collaboration-with-forks/)

**Comment by stropho** [comment]
Excellent. The checkbox has been checked 
image

**Comment by patrik-piskay** [comment]
And it's live in `v5.0.0` 🎉  thanks for your work on this @stropho!

**Comment by patrik-piskay** [comment]
> One thing I noticed, though, is that once RTM renders an empty string (a long word with "word" tokens, or a long "Atom"), it renders an empty string (good!), but when resizing its box it never recovers - it always renders an empty string. Not specific to this PR as I was able to repro with word tokens, so I'll create an issue and will try to fix it separately.

I actually fixed this behaviour, but decided to not render an empty string with `...` if there is no space. Instead, at least one atom (character/word/`<Atom />`) will be rendered (even if breaking the allowed number of lines) so that the ellipsis makes sense. It's a UI problem if it ever gets to a point where there is not enough space for at least one atom

**Pull request merged**